### ISSUE TITLE
File-choosers can appear to bypass filesystem restrictions

### ISSUE THREAD

**Issue by sillitatti** [flatpak/flatpak#4991]
With GTK apps the restrictions are applied properly but Qt apps are able to access the full host file system via the file picker regardless of sandbox settings.

Is this a known issue? I tried to search but couldn't find anything related. Can anything be done about it? 

Running on Fedora 35 KDE/X11.

**Comment by smcv** [comment]
> Qt apps are able to access the full host file system via the file picker

Please describe the exact situation you are encountering, what you thought should happen, and what actually happened?

What is meant to happen is:

* `--filesystem` controls what the app can read/write *without* user interaction. For example, if an app has `--filesystem=xdg-download/Firefox` then it can download files into `~/Downloads/Firefox` or similar without waiting for user interaction with a file chooser.
* Additionally, if the app communicates with xdg-desktop-portal to show a file chooser (a typical GUI app's "File -> Open" or "File -> Save As..."-style UI), then it is given access via the documents portal to whichever file the user chose in the interactive file-chooser. As a user, if you don't want the app to access a particular file, don't choose it (cancel the file chooser dialog instead).

**Comment by refi64** [comment]
It's not a security issue because the file dialog is actually run on the host system, and only the selected files/folders are exposed to the application. Because the dialog isn't in the app's territory, it can't be controlled by the app at all outside of some very basic functionality (e.g. setting the file filters).

I'm guessing the GTK apps you tested didn't actually support the file dialog portal, so that dialog *was* in fact running inside the Flatpak itself.

**Comment by smcv** [comment]
I wonder whether it'd make this more obvious to people if the portal file dialogs explicitly said something about giving access? Something like adding the last line to this (I'm basing this ASCII-art on the GTK 4 dialog from xdg-desktop-portal-gnome, but any implementation could do something similar):

```
[ Cancel ]                 Select an Image              [ Open ]
----------------------------------------------------------------
other standard file-chooser stuff goes here
----------------------------------------------------------------
The application will be given access to the selected file.
```

On one hand, the whole point of the portals is that where possible, we're trying to have the overall system of apps, Flatpak, portals and the host system automatically and transparently do the right thing, without users needing to think about permissions, restrictions or the existence of portals; so having extra noise in the file-chooser is distracting and makes the overall system look less well-integrated.

But on the other hand, the file-chooser is perhaps doing this job too well - it integrates so well with the requesting app that users semi-regularly think they've found a serious problem!

Doing that would also give a hint about access to the selected file being permanent (the read or read/write permission is saved and the app can continue to access that specific file in future), which is not necessarily obvious to everyone. (The reason why this needs to happen is that if it didn't, then File -> Save, File -> Revert Changes and File -> Open Recent -> foo.doc wouldn't work you'd expect.)

**Comment by smcv** [comment]
> I'm guessing the GTK apps you tested didn't actually support the file dialog portal, so that dialog was in fact running inside the Flatpak itself.

If you don't have any GTK apps installed and you want to verify this, GNOME Recipes is a useful simple example (click on New Recipe, and then on the "+" below the camera icon to load an image file).

**Comment by sillitatti** [comment]
> Doing that would also give a hint about access to the selected file being permanent (the read or read/write permission is saved and the app can continue to access that specific file in future), which is not necessarily obvious to everyone. (The reason why this needs to happen is that if it didn't, then File -> Save, File -> Revert Changes and File -> Open Recent -> foo.doc wouldn't work you'd expect.)

Indeed, it made me wonder when I noticed that apps were able to directly load previously edited files with the same permissions as the file picker even after closing and reopening the app.
Having those automatically saved permissions listed and editable somewhere in a human-readable format might be useful.

Thanks for all the clarifications.
